Added TestCaseInfoHasher and tests.

danifeist opened the pull request:
This PR tries to fix #2304 "Test case name hashing should consider tags and class name too".
The goal is to "allow test cases with same name, but different tags, or same names and tags but different class name".

See the following code example that describes what is wanted:

```cpp
Catch::StringRef className1{std::string("C1")};
Catch::NameAndTags nameAndTag1{std::string("N1"), std::string("[T1]")};

Catch::StringRef className2{std::string("C2")};
Catch::NameAndTags nameAndTag2{std::string("N2"), std::string("[T2]")};
```

this set of test cases satisfies the wanted unique test case invariant as soon as:

```
if (N1 == N2 || C1 == C2) then [T1] != [T2]
AND
if (C1 == C2 || [T1] == [T2]) then N1 != N2
```

Before this PR, the code behaves as follows:
When `TestRunOrder::Randomized` is set, `getAllTestsSorted()` does the following:
1) Use `TestHasher` hash the class name of a `TestCaseInfo` (using the FNV-1a algorithm) and push them to a vector.
2) Sort the vector by the hashes, on equality, compare name, if equal compare class name, if also equal compare tags.

After this PR, the code behaves as follows:
When `TestRunOrder::Randomized` is set, `getAllTestsSorted()` does the following:
1) Use `TestCaseInfoHasher` to generate hashes that considers class name, name and tag and push them to a vector.
2) Sorting the vector by the hashes. When the unique test case invariant described above is satisfied, the hashes will be unique.

This is my first contribution to Catch2, hope i went in the right direction and would be happy about feedback, comments, etc.

danifeist committed:
Added TestCaseInfoHasher and tests.

horenmar commented on src/catch2/internal/catch_test_case_registry_impl.cpp:
```diff
@@ -79,11 +55,7 @@ namespace {
-                                      rhs.second.getTestCaseInfo();
-                           }
-                           return lhs.first < rhs.first;
+                              return lhs.first <= rhs.first;
```
This is wrong, the original code was correct -> the hashes can collide even if the inputs are different, so we want to use the `TestCaseInfo` comparison as the tiebreaker to ensure consistent order.

horenmar commented on tests/SelfTest/UsageTests/RandomlyOrderedTestCases.tests.cpp:
```diff
@@ -0,0 +1,50 @@
+
+#include <iostream>
+
+static constexpr Catch::SourceLineInfo dummySourceLineInfo = CATCH_INTERNAL_LINEINFO;
+
+TEST_CASE( "Set TestRunOrder of command line parser to Randomized and register tests in registry." ) {
+    Catch::ConfigData configData;
+    auto cli = Catch::makeCommandLineParser(configData);
+    CHECK(cli.parse({"test", "--order", "random"}));
+    REQUIRE(configData.runOrder == Catch::TestRunOrder::Randomized);
+
+    Catch::StringRef className1{std::string("name")};
+    Catch::NameAndTags nameAndTag1{std::string("nameA"), std::string("[tagA]")};
+    auto ptrToTestCaseInfo1 = Catch::Detail::make_unique<Catch::TestCaseInfo>(className1, nameAndTag1, dummySourceLineInfo);
+    auto ptrToTestInvoker1 = Catch::Detail::make_unique<Catch::TestInvokerAsFunction>( []{std::cout << "f1\n";} );
+
+    Catch::StringRef className2{std::string("nameX")};
+    Catch::NameAndTags nameAndTag2{std::string("nameA"), std::string("[tagA]")};
+
+    auto ptrToTestCaseInfo2 = Catch::Detail::make_unique<Catch::TestCaseInfo>(className2, nameAndTag2, dummySourceLineInfo);
+    auto ptrToTestInvoker2 = Catch::Detail::make_unique<Catch::TestInvokerAsFunction>( []{std::cout << "f2\n";} );
+
+    Catch::StringRef className3{std::string("name")};
+    Catch::NameAndTags nameAndTag3{std::string("nameA"), std::string("[tagB]")};
+
+    auto ptrToTestCaseInfo3 = Catch::Detail::make_unique<Catch::TestCaseInfo>(className3, nameAndTag3, dummySourceLineInfo);
+    auto ptrToTestInvoker3 = Catch::Detail::make_unique<Catch::TestInvokerAsFunction>( []{std::cout << "f3\n";} );
+
+    Catch::TestRegistry testRegistry;
+    testRegistry.registerTest(std::move(ptrToTestCaseInfo1), std::move(ptrToTestInvoker1));
+    testRegistry.registerTest(std::move(ptrToTestCaseInfo2), std::move(ptrToTestInvoker2));
+
+    Catch::Config config(configData);
+    const auto& allTests = testRegistry.getAllTestsSorted(config);
+
+    for (const auto& test : allTests)
+    {
+        test.invoke();
+    }
+}
```
No, remove all of this.

I am going to write up more about tests later, but trying to do this inside a unit test like this doesn't make sense.

horenmar commented on src/catch2/internal/catch_test_case_info_hasher.hpp:
```diff
@@ -0,0 +1,20 @@
+class TestCaseInfoHasher {
+public:
+    using hash_t = std::uint64_t;
+    TestCaseInfoHasher() = default;
+    ~TestCaseInfoHasher() = default;
```
These can be removed completely.

```suggestion
```

horenmar commented on src/catch2/internal/catch_test_case_info_hasher.hpp:
```diff
@@ -0,0 +1,20 @@
+#ifndef CATCH_TEST_CASE_INFO_HASHER_HPP_INCLUDED
+#define CATCH_TEST_CASE_INFO_HASHER_HPP_INCLUDED
+
+#include <catch2/catch_test_case_info.hpp>
```
Move this include down into the .cpp file and only fwd declare `TestCaseInfo` in the header.

horenmar commented on src/catch2/internal/catch_test_case_info_hasher.cpp:
```diff
@@ -0,0 +1,29 @@
+#include <catch2/internal/catch_test_case_info_hasher.hpp>
+
+#include <iostream>
```
This looks like a debugging leftover

danifeist committed:
Adaptions after code review:

- Kept original TestCaseInfo comparison logic so that hashes can collide even if the inputs are different
- Removed RandomlyOrderedTestCases.tests.cpp
- Updated ApprovalTests
- Minor cleanup

horenmar commented on src/catch2/internal/catch_test_case_registry_impl.cpp:
```diff
@@ -66,9 +41,10 @@ namespace {
 
-            TestHasher h{ config.rngSeed() };
+            //TestHasher h{ config.rngSeed() };
+            TestCaseInfoHasher h;
```
Where did the seed handling go?

horenmar commented on src/catch2/internal/catch_test_case_registry_impl.cpp:
```diff
@@ -79,7 +55,7 @@ namespace {
                        indexed_tests.end(),
                        []( TestWithHash const& lhs, TestWithHash const& rhs ) {
-                           if ( lhs.first == rhs.first ) {
+                          if ( lhs.first == rhs.first ) {
```
whitespace fix

```suggestion
                           if ( lhs.first == rhs.first ) {
```

horenmar committed:
Update src/catch2/internal/catch_test_case_registry_impl.cpp

horenmar commented on tests/SelfTest/IntrospectiveTests/TestCaseInfoHasher.tests.cpp:
```diff
@@ -0,0 +1,41 @@
+        Catch::TestCaseInfoHasher h;
+        CHECK(h(testCase1) != h(testCase2));
+    }
+}
```
This looks pretty reasonable.

You should also add a test that the hasher can be seeded (and that the same `TestCaseInfo` produces different results for differently seeded hashers).

I would also nitpick different section names, but those are just nitpicks and these are okay.

danifeist replied:
When i understand you right it shall be possible to seed or not seed the TestCaseInfoHasher, that is how the last commit d8e6863 implements it.

The specification of FNV-1a https://datatracker.ietf.org/doc/html/draft-eastlake-fnv-17.html states that "almost any offset_basis will serve so long as it is non-zero". Currently one can pass a seed of zero to TestCaseInfoHashers constructor which we could prevent with for example a fallback but i don't think that's really an issue.

horenmar replied:
Right, I wasn't clear enough originally.

The original code used a known good seed for the hash, but
then mixed in seed from Catch2 to randomize the result. Let's call these
the hash-seed (the `hash_t hash = 14695981039346656037u` in the original
code), and Catch2-seed (what Catch2 passes into the hasher from user config).

The reason it is done like this is that the test case hashing is used
to implement `--order rand` in a stable way. The test cases are hashed,
the hashes are sorted, and the resulting order is taken as the order
of test cases to be run.

So what should be done is that the algorithm keeps the known good hash-seed
(ideally the same one), and mixes in the Catch2-seed at the end, the way
the original implementation used to do. This includes the shortened
multiplication, as it improves the scattering of similar test cases.

danifeist replied:
Thank you for the detailed explanation, i read it yesterday and was about to commit almost the same as you did today but hesitated and felt i have to think a bit about the constructor-with-no-parameter case that you removed which is fine. :-)
The latest commit adapts the unit tests and the return value type.

horenmar commented:
Thanks for the effort so far, only minor changes are needed to merge this.

danifeist committed:
Further adaptions after code review:
- Added seed handling to TestCaseInfoHasher
- Enhanced unit tests with seeded hashers
- Updated ApprovalTests

horenmar commented on src/catch2/internal/catch_test_case_info_hasher.cpp:
```diff
@@ -0,0 +1,30 @@
+    TestCaseInfoHasher::operator()( TestCaseInfo const& t ) const {
+        // FNV-1a hash algorithm that is designed for uniqueness:
+        const hash_t prime = 1099511628211u;
+        hash_t hash = m_seed;
```
```suggestion
        hash_t hash = 14695981039346656037u;
```

horenmar commented on src/catch2/internal/catch_test_case_info_hasher.cpp:
```diff
@@ -0,0 +1,30 @@
+                hash *= prime;
+            }
+        }
+        return hash;
```
```suggestion
        hash ^= m_seed;
        hash *= prime;
        const uint32_t low{ static_cast<uint32_t>(hash) };
        const uint32_t high{ static_cast<uint32_t>(hash >> 32) };
        return low * high;
```

horenmar commented on src/catch2/internal/catch_test_case_info_hasher.cpp:
```diff
@@ -0,0 +1,30 @@
+namespace Catch {
+    TestCaseInfoHasher::TestCaseInfoHasher( hash_t seed ): m_seed( seed ) {}
+
+    TestCaseInfoHasher::TestCaseInfoHasher(): m_seed( 14695981039346656037u ) {}
```
```suggestion
```

horenmar commented on src/catch2/internal/catch_test_case_info_hasher.hpp:
```diff
@@ -0,0 +1,23 @@
+    public:
+        using hash_t = std::uint64_t;
+        TestCaseInfoHasher( hash_t seed );
+        TestCaseInfoHasher();
```
```suggestion
```

horenmar committed:
fixup handling of Catch2-provided seed

danifeist committed:
Change of tests after Catch2-provided seed was added

- TestCaseInfoHasher satisfies clang-format rules
- Updated ApprovalTests
